app/common/lib/ledgerUtil.js

```diff
@@ -241,35 +263,184 @@ const getMediaData = (xhr, type) => {
+
+  if (!parsedUrl || !query) {
+    return null
+  }
+
   switch (type) {
     case ledgerMediaProviders.YOUTUBE:
       {
-        const parsedUrl = urlParse(xhr)
-        let query = null
+        result = queryString.parse(parsedUrl.query)
+        break
+      }
+    case ledgerMediaProviders.TWITCH:
+      {
+        result = queryString.parse(parsedUrl.query)
+        if (!result.data) {
+          result = null
+          break
+        }
+
+        let obj = Buffer.from(result.data, 'base64').toString('utf8')
+        if (obj == null) {
+          result = null
+          break
+        }
 
-        if (parsedUrl && parsedUrl.query) {
-          query = queryString.parse(parsedUrl.query)
+        try {
+          result = JSON.parse(obj)
+        } catch (error) {
+          result = null
+          console.error(error.toString())
         }
-        result = query
         break
       }
   }
 
   return result
 }
 
-const getMediaDuration = (data, type) => {
+const getMediaDuration = (state, data, mediaKey, type) => {
   let duration = 0
+
+  if (data == null) {
+    return duration
+  }
+
   switch (type) {
-    case ledgerMediaProviders.YOUTUBE: {
-      duration = getYouTubeDuration(data)
-      break
-    }
+    case ledgerMediaProviders.YOUTUBE:
+      {
+        duration = getYouTubeDuration(data)
+        break
+      }
+    case ledgerMediaProviders.TWITCH:
+      {
+        duration = getTwitchDuration(state, data, mediaKey)
+        break
+      }
   }
 
   return duration
 }
 
+const generateMediaCacheData = (parsed, type) => {
+  let data = Immutable.Map()
+
+  if (parsed == null) {
+    return data
+  }
+
+  switch (type) {
+    case ledgerMediaProviders.TWITCH:
+      {
+        data = generateTwitchCacheData(parsed)
+        break
+      }
+  }
+
+  return data
+}
+
+const generateTwitchCacheData = (parsed) => {
+  if (parsed == null) {
+    return Immutable.Map()
+  }
+
+  if (parsed.properties) {
+    return Immutable.fromJS({
+      event: parsed.event,
+      time: parsed.properties.time
+    })
+  }
+
+  return Immutable.fromJS({
+    event: parsed.event
+  })
+}
+
+const getMediaFavicon = (providerName) => {
+  let image = null
+
+  if (!providerName) {
+    return image
+  }
+
+  providerName = providerName.toLowerCase()
+
+  switch (providerName) {
+    case ledgerMediaProviders.YOUTUBE:
+      {
+        image = require('../../../img/mediaProviders/youtube.png')
+        break
+      }
+    case ledgerMediaProviders.TWITCH:
+      {
+        image = require('../../../img/mediaProviders/twitch.svg')
+        break
+      }
+  }
+
+  return image
+}
+
+const getTwitchDuration = (state, data, mediaKey) => {
+  if (data == null || mediaKey == null) {
+    return 0
+  }
+
+  const previousData = ledgerVideoCache.getDataByVideoId(state, mediaKey)
+
+  if (previousData.isEmpty() && data.event === 'video-play') {
+    return milliseconds.second * 10
+  }
+
+  if (!data.properties) {
```

Comment:
should this be moved up to the initial check? ex:
`if (data == null || mediaKey == null || !data.properties) {`

Reply:
we can, done